Some cleanup and fixes found using perl5 plugin for Intellij IDEA

hurricup opened the pull request:
I checked source with my plugin and found few problems of different severity.
Each commit is atomic changed and may be accepted or rejected.
There are more fixes which may be done there. 
Also, i'd like to re-format file with some consistent code style.

smxi commented:
I don't accept patches based on automated tools. Those tools are not 'smart' enough to 'know' what I want to be happening.

I will take a look through those patches to see if there's anything that should actually be 'fixed', but every time I've  run one of those tools the results were so wrong and absurd that I stopped paying any attention to them.

I will in particular check the unused variables, because I do frequently do cleanups of old logic and forget to remove the variables, but sometimes I have old variables that I actually want there even though unused at the present time.

Some of them are just arbitrary however, in my coding practice, I NEVER use these types of code check tools, they are not valuable in my experience, and can lead to wrong code since no tool made by humans has yet been able to reach the point of 'understanding' what it is seeing.

thanks for the thought however.

hurricup commented:
All changes made manually. Tool just found problems. Not fixed them. But do as you feel right. And you need to check this anyway.

smxi commented:
Yes, I will, inxi hasn't seen a check for stray data for a while, sometimes I will go through a chunk and clean up stuff that has changed or is wrong, usually in the help or man page that tends to happen a lot since I don't actually proofread them totally, I usually just check the part I'm working on, but now and then I will go through the entire thing, so this is not a bad idea to check the stuff.

The problem with trying to do a patch to inxi is that the lines won't match up, I just did a bunch of real changes internally for example in pinxi to deal with an issue, so the line numbers no longer match up, which makes manual checks more difficult, but not that bad since the lines have probably only changed by 50 or so as it currently is.

It is however funny that two of those check diffs actually contradicted each other, but I'll double check everything, appreciate you taking the time to do this.

hurricup commented:
Just a thought: i'd considered a test for such complicated project. With mocking environment. Just need to figure out architecture. Probably using some proxy module for data collection, which can be substituted in tests.

smxi commented:
Tests are very difficult because a big chunk of data comes from /sys. the debugger handles that, but can't duplicate the actual /sys structure in terms of individual files, but the debugger (--debug xx) does collect in file form most of the data used.

tests are only of limited utility however because the number one cause for real bugs in inxi is unexpected or non standard data or data structures or data syntax.

because of the complexity, there are many ways to inject data into inxi, particularly for very hard to debug stuff where the literal user data with the issue or failure or bug is used internally, but that's a manual operation always, there are also many debugging flags to be used for specific and often very prone to failure components, but I would not personally spend too much time with trying to make some abstraction layer in terms of how inxi processes data, it would be way too much work, and due to the extremely specific nature of debugging inxi, very unlikely to actually work in the way needed, sometimes manual insertion of key data in file format is just the easiest way to do debugging, sad to say.

hurricup commented:
I'm not really familiar with things it does inside, what data it reads and so on, but from development perspective I believe that it would be nice to have tests running against different data you've obtained from users (probably anonymized if needed) and  comparison with some serialized output.

This could help you to maintain and could help others to contribute. At least to check that nothing broken by changes (with some precision). 

Also, if I wanna fix a bug on my system, I should be able to dump data + result. fix something, run again and see a diff. 
But yes, I don't really know how hard this might be. Just a thought. Again  :)

Thanks anyway  for your time :D

smxi commented:
￼Fixed unreachable code lines: 
done, those were good fixes, lots of oversights there, copy paste errors, etc.

Unused variables:
 takes too much work to double check each item, the starting $t1/$t2 are debuggers that can be uncommented, it's very unwise to remove things just because they do not appear when you ignore comments, just search the file and see if they are used, and if they are, particularly for debuggers, they aren't actually things that should be removed, I'll go through that later when I feel like spending the time. This one I have to be super careful with, for example, while @sysctl_disks may appear to be not used, it's actually a note to myself to one day handle and add sysctl_disks though it's not there now. If you look above that, you will see a bunch of other sysctl items, that are used, this was one I was thinking about using, and will maybe one day use, if it's there to remind me I had that idea. I use these types  of notes internally to remind myself that there's a feature I was going to do one day but never got around to, so I think for the variable one I am not going to do it because I have to check each and every one to make sure it's actually correct.

For instance:
```
my @temp = @$three;
# print Data::Dumper::Dumper \@$three;
```
removing that my @temp = @$three; is wrong, the glitch was not commenting out that line to fit along with the commented out print debugger there, that's the problem with these automated code checks, they don't understand a single line of code or context, and are ignoring commented out data.  that makes them VERY dangerous to use, but I will go through those and correct the stuff where I understand what I'm doing for sure, a lot of these will be debuggers, for instance, when I give a user directions to comment out a certain debugger, which then uses $t1 and $t2, those have to already be declared, I can't have users trying to figure out anything beyond the simple uncomment one line I will sometimes ask them to do. So $t1 and $t2 become used the instant you uncomment a debugger item, that is. But I can't ask a user to declare those too, that's too many steps for them.

Changed discouraged fancy object calls to one with arrow: 
this one I'm not sure about, are there any perl 5.008, 5.010, 5.012, and so on, issues with using this method? 
```
import File::Copy;
File::Copy->import;
```
where is it discouraged?, I believe I was following some standard, and encouraged, documented method when I created these. How do we know that a module supports the import method? I'm not a perl expert, but I lean always towards using the stuff as it's been documented, and only updating the code when it's definitely something I'm sure about.  And with methods I have tested on perl 5.008 with no issues, I'm always leery of altering some core things in inxi that might appear to be modern and correct but which might break perl 5.008 or other later variants, which is the perl inxi is designed to run on, basically the idea is for the main rpm and deb systems that might be really old, like a 10 year old server, inxi is supposed to 'just work', always, without exception.

Cleanup removed redundant namespace qualifiers: 
while I did this one, it actually slightly damages the maintainability since I can safely do global search and replaces for main::key( but I cannot safely do global search and replaces for key(
However, I just did a big change internally and do not expect to need to modify these again for a while, but technically I should probably have left all those key( as main::key because it's a much safer search/replace, which is actually why they were there in the first place. I believe, though I'm not certain, that in this case, main::key is the same exact thing to perl as key in this context, so I'm not sure this is a real style rule or just someone's preference.

Bugfix: added namespace qualifier for main:: subs: 
this was a bug, thanks, although a bug that would never show for any normal user because logging has not started at that point, the only way to start it there is manually by forcing it in the first set of debuggers manually in sub main. Actually it's never shown for any user, including me, not sure why, I've run this with logging lots of times, but it's good to have it fixed anyway.

smxi commented:
thanks for taking the time to look, but be VERY careful with those automated code checkers, first, they are sometimes wrong, plain and simple, and simply reflect one single person's idea of how they think perl code should look, and that idea might actually be technically wrong. Early on I did a LOT of optimization testing to make sure to avoid radically bad and unefficient practices, and I did use these code checkers a few times but I found their results so bad that it would have been actually worse using them than not using them, though they are good for certain things, like finding lines that will never execute, and the bugs you found, so I don't say they are useless, but NEVER EVER TRUST THEM!!! Use them as guides, but to me, they cost almost more time t use than to not use.

Debugging inxi is very difficult, I think I would suggest you see if you can figure out how to debug it using its built in tools, which are partially documented, though big stress on partially since in general I'm the only person who ever sees or uses them, here: 
inxi-perl/docs/inxi-values.txt
note that these docs are often totally out of date, they are a low priority for me, and you can roughly see which I actually use and which I made once, and sometimes some of them are just code snippets I don't want to lose, it varies, but the values shows some of the debuggers and switches, for instance, I can roughly emulate a bsd but I have to setup a bunch of code imports, it's a real pain in the butt to do, I rarely do that, I will instead only do a single feature test for a bsd for example.

hurricup commented:
This is not a code checker, this is more like [perl IDE](https://github.com/Camelcade/Perl5-IDEA). And I am the author. Working with real projects is my dogfooding, helps to invent new features and check if existing works right, because i'm not programming in perl these days. For example, to clean up this project, i've implemented proper handling of evaled variables in proper context. 

It is not plain, simple and been in development for 5.5 years now. Of course it can't be 100% right, because it's a perl. Again, it's purpose is to help and save time, not doing work for developer. And no, it doesn't take more time than manual work. And i'm not trusting them.

Regarding changes:
1. unused variables. 
- I've checked comments. It was only line using $t1 and i've removed it as well. Look a not very good practice to me. Debugging should be enabled with command line argument, not commenting code.
- keeping ideas in unused variables is really strange. Isn't tracker and issues created exactly for this?
- again: no automated tool removed anything. Everything been done manually. it feels like you are afraid of tooling for some reason, but this is really strange. They are tools, not terminators. People are using them. Tool just shows that variable is unused. It should be commented out or removed. I prefer the last one, because I strongly believe that switching to/from debugging mode by changing sources is a really bad practice. It should be achieved with some command line argument, and source code should stay intact. 
2. I'm not aware about any problems using `->`. IRR this `fancy` call is just a discouraged alternative, made for calls like `new Foo::Bar`. See https://perldoc.perl.org/perlobj.html (indirect object syntax)
```
Outside of the file handle case, use of this syntax is discouraged as it can confuse the Perl interpreter. See below for more details.
```
3. Redundant namesapaces. yes, it works and it is correct. Just redundant. Works without it properly.

hurricup commented:
And sorry, i've missed branch thing in readme :(

smxi commented:
LOL, 'strange' is a relative term. Basically I do all the dev of inxi, with a tiny fractional percent done over the years by a few other people, but they never stay so I don't really care about their needs, since they are not here, and never have been, I do all the work, almost all, so this stuff reflects how I want it to be. There's no 'team', I don't have to make my code worse to make it more usable for others, it just can be what it is. 

Be careful of abstract 'rules' like debugging should happen in x or y fashion, that's not really how it works in my opinion, debugging should be done in the way that works, and which experience shows it should be done. In other words, it should work for ME, since I'm the primary user. While the case of a theoretical developer who theoretically gets interested in inxi/pinxi for a few months before losing interest and vanishing might be of interest theoretically, it's of essentially zero interest to me. I've gotten stuck with maintaining some internal data tables that have to be empirically updated constantly because in one case, the guy who asked for the feature said he'd maintain it in terms of research and updates for the future, he never even stuck around long enough to do the first revision, so I don't do anything for theoretical cases that do not exist, and which actually make things harder for me in the real world of real coding.

Sure, if you are working in an enterprise situation you have to do a lot of things to make the code workable for an overall mediocre developer pool, and keep things in a certain way, but that has NOTHING to do with quality or ease of development, that's just something larger projects need, though I challenge anyone to actually debug a kernel issue in any reasonable way using command line arguments, lol, nor most any other real project, that just doesn't happen. That's a rule that would only apply to a super static project done in a static way, one that probably moves very slowly, where you can assign to drones  certain tedious steps. Projects like that would probably take 10x longer to do a new feature like inxi introduces in a few hours, lol, if they could ever even achieve that result.

I"ll read up on the indirect object syntax) and update it, but I try to never do things that I don't personally understand, but once I understand them I'll use them. Perl is obviously filled with things I don't understand, lol. 

But I would caution making up rules like saying debugging should be done via command line, that's simply not possible in the case of inxi, it would take HUGE work to do it, and would result in zero benefit to me. Some debugging happens with command line, with one exception, all user actions for debugging are done in the case of command line (and in my opinion, done better than a huge percentage of other projects, in some cases, MUCH better, that's by necessity, because I need user triggered debugging events to be super simple and easy for users so I can actually get the required data).

I may add back in the main::key because there's a maintenance issue there that may bite me in the future, there's also just a real world issue that it's easiest to be able to copy paste that stuff that generates keys anywhere. 

I'm not really afraid of stuff like tooling, I just don't need it, nor does it benefit me, and every case I've seen of such things offering recommendations just hasn't been worth the time it would require to check each thing. If inxi were a tool that ran as a daemon continuously I'd probably treat certain things slightly differently, but not that differently.

Again, the complexity of what inxi has to deal with internally makes the notion that each and ever feature that can be debugged or needs to be checked or use alternate code would need its own little logic block to do so would be almost clinically  insane to use and code, that would severely diminish my ability to actually develop and code inxi, it would that is be a negative, not a positive, it would add complexity and remove flexibility, in other words, it would be a regression, and a quite serious one.

Add to this that as far as I know, there is exactly one user in the world who actually knows how to even use the command line debuggers that exist, me. I'm certainly the only one I'm aware of who has ever said anything about them, nor are they easy to use correctly. 

I find an oddity, there's this Perl saying, “There's More Than One Way To Do It!”, but I've noticed over the years that while Perl people will paste that acronym all the time, they don't actually believe it or follow it, for instance, the suggestions re 'rules' to follow here ignore the way I do it, which is the way I like it, which is based on doing it for over 10 years in inxi, and I even had the benefit of doing the rewrite to perl 2 or so years ago, which let me do it even more the way I want to do it, which means, perl inxi is very close to exactly the way I want to do perl and inxi, in fact, I liked this way so much I extended it to my other perl program, acxi over the last year.

This is also why I don't in general ask perl people for code reviews or anything like it, I became aware quite early on when I was upping my perl level to do inxi perl, that while they say TMTOWTDI all the time, like a mantra, they really don't follow that notion, but like to make up stuff about how to do it that ignores a lot of coding reality in my experience. This is why I've literally never asked perl people for a code review, nor do I ever mention acxi or inxi if I have an issue I can't resolve, I only post the literal code I'm having an issue with if I can reduce it to a few lines of logic.

But I really suggest being careful of making up rules that can end up being negative in terms of realworld fast agile development, which is essentially what I do, though I did it long before that term became hip, to me it's the only way to do real code in a realworld environment, at least it is the only way for me to do it.

hurricup commented:
Of course you as a single developer should do as you want. Just saying that current approach makes your project almost impossible to contribute to. Probably this doesn't change anything, but we'll never know :)

smxi commented:
You're focusing on an issue which has never existed. In other words, there have never been real contributors. There were people who worked on a few components, but that's it. I can tell you one thing, for a time, I allowed contributors to sgfxi and smxi, from the distro we were in, and I found the results so BAD that it was a wakeup call, most people can't code, and since I code for a living, the last thing I want to deal with is bad code from bad programmers, so I consider the bar to entry of inxi as a feature, not a bug, I don't want code from bad programmers, it's no fun to deal with, and if I'm not getting paid to deal with it, why on earth would I do something for free that is no fun?

Other rules inxi/pinxi violates is: 
* code must work in vi, despite vi being a 30, 40 year old tool. 
* code must be 80 columns wide, only, despite 80 column monitors not even existing as a normal thing for over 25 years.
* code must be split into tiny annoying chunks then reassembled into one unit despite it being far easier, and way better for end users, to have it be in one single file.
* code must be ok for users unwilling to learn how to use code folding, or for users who are incapable of doing code folding due to bad tools or bad technique.

Again, while what you say would have been _faintly_ interesting about 10 or 12 years ago, when inxi launched, sadly the reality of free software projects is that the people who do the work should be able to determine how that work happens. Since nobody else did the work, they don't get any say, since they don't exist. 

However, with this said, let's say that one purpose of inxi is to weed out bad or drone or mediocre programmers. In other words, let's say that inxi as my project has zero interest in catering to or supporting anyone not capable of dealing with its physical layout.. From what you are saying, it sounds then like I did it right. 

Note further, as the years pass, I write inxi to be developed and optimized for its programmer, me. Since I can't be anyone else than me, would just not work out, who else then should I be writing it for? Obviously I want it to work for me as best as it can, I want it to take the least work possible to maintain and develop and debug, not the most, so that's what it does.

I don't exclude that one day some smart programmer who uses good tools, not legacy stuff, and who can follow folded logic, use the search features, might take interest, but the reality is, inxi is really complicated, it's far far far more complicated internally than most people will ever realize, because the horrible mess that is realworld data in real systems is incredibly messy.

Anyway, I've never once seen these mystical programmers who I should make inxi work for, lol, not once, not in any free software poject. I did now and then find a smart guy who was able to figure out stuff, they contributed, they were able to work with inxi, no real issues, found a few optimizations, at which point, I realized, ah!!! that's who I want to cater to!!!

However, your belief that this project is impossible to contribute to is wrong, it's just got a high bar, you have to understand the logic, it's not about debuggers or structures, that's a tiny fraction of the actual problem inxi faces, the real problem is in the logic, and that logic is what programmers need to grasp, and if they don't want to grasp it or understand it, how could they ever possibly submit a useful patch beyond the most trivial?

smxi commented:
Just a quick lesson in how inxi actually works:

* inxi has a set of black boxes, these run core things like output etc. users NEVER would touch those, ever, they were very difficult to test and develop, and should never be touched by anyone. They have an interface, which is all the user needs to know to work with them.

* inxi has a set of tools to do repeated tasks, small tools, the user has to learn these tools so that they don't manually rewrite them each use of the logic in the working code.

* inxi has data generators, that generate the actual lines and sub items in the lines, like Info and System. In most of them, they generate data structures that will be sent to the output handlers

* last step is to actually fire the events that will fire the data generators which will then return the data structures which will then be sent to the printers or output features.

In other words, in all practical senses, there is only one component a user would ever touch, the main data generator blocks, but only after learning the tools to avoid redundant code, and learning how the output generator needs the data.

This is the logic, it's irrelevant, totally irrelevant, having command line debuggers, but there are, again, command line debuggers, there are a LOT of them, for features that need it, for the part of the logic that needs it.

Once  a coder shows up who demonstrates that they have learned the logic, obviously I'd welcome their contributions, but I'd NEVER try to make something for a coder who does not exist, all internal features of inxi are made for coders who do exist. With this said, I often get issues where the person basically figured out what was wrong internally, and had the right fix for it.

smxi commented:
By the way, if somehow you want to actually learn the logic of inxi, and spend the time, feel free to poke around with it, but remember, it's the logic that is the challenge in inxi, not debuggers, and the more you poke around, the more lines of commented out debuggers you'll see, and sometimes, command line triggered ones, but not usually, but it's the logic that is the real challenge here, NOT formal or stylistic elements. Also realize, current inxi is the result of about 10 years of bash/gawk inxi, where I built up a set of things that I wanted to do but which I could not do, so the current perl inxi is in fact the solution to the real issues I found doing the old inxi, so basically, it all happens the way it does for a reason, and that reason is a process that began 12 years or so ago, and it heavily depends on a strong awareness of what garbage user data really is, and how to deal with it, in a way that usually works for most users. The perfect is the enemy of the quite good, that is, in this case. Or not. 

Note that old bash/gawk inxi was basically impossible to work on, it was nearly impossible for me, which is why I avoided it as much as possible, and tended to do work on it only in large time slots since it was too hard to get into the bash/gawk mindset. The new inxi seems to benefit from my software experience because basically all the issues I wanted to fix were fixed, and all the flexibility I wanted is here, something I realize every time I do a huge internal extension of an existing featureset, the core syntax and methods of inxi actually work very well, exactly as I intended them to do, so I definitely would say, if you think inxi and its style is wrong, it's wrong for you, but it's right for inxi and its users, because it works SUPER well, like, way way better than I dreamed when I did the perl rewrite, I knew roughly what I needed, but I didn't realize I'd largely succeeded. This became evident recently when I did the -y1 key: value pair indented output option, and it took almost no time to do, it did take time to do the switchover because the logic had to be hand done where search and replace failed, then manually checked, but that was the last big proof I needed to see that my dream / intention was a success, new inxi works super well, and has never hit me with any obstacle that its internal logic can't handle.

hurricup commented:
I am. And I'll probably will play with it to learn and test some things. But i'm too spoiled with well-organized and well-tested projects, not just good working for years.  So for me it requires additional efforts without any benefits. Investigating, learning and fixing is cool. Walking through the minefield with a stick in some specific manner just because someone finds this most comfortable personally - not so much.

smxi commented:
That's fine, won't change anything I do. There's no minefield here however, it is internally really well organized, but it requires doing the work, and being able to retain conceptual information in your brain beyond the trivial, which is really the same as any actual software out there. I'm somewhat mystified by the projects you like since I've never seen anything like that, my experience with linux kernel debugging was their debuggers were truly dismal, a view during my issue shared by andrew morton if I remember right, and the process to generate that stuff was absurdly complex, so you must be talking about some other projects. Maybe you are selecting ones that happen to fit how you like things done, and ignoring the ones that don't do that? I really don't know, all I know is that if you use the command line debuggers, you will get several orders of magnitude more debugging data than any comparable tool offers, and the user debuggers are a core part of inxi since most new features cannot be developed without those debuggers.

With this in mind, you may simply be using words in a totally different way than I use them, but I've  never seen anything even remotely like what you are talking about in some of the bigger projects in the world whose code I'm familiar with, stuff several orders of magnitude bigger than inxi, so is it possible that because you want to see certain things done in a certain way, you maybe confuse this with 'good practice' or whatever?

I am sort of genuinely puzzled, I'm obviously not trying to write newbie friendly or generic programmer friendly enterprise type code here, I'm trying to write code that works and can be developed VERY quickly, with new features rolling out in a way that respects my finite time on earth as a developer. Given that inxi appears to continue to be reasonably unique and continues to expand its userbase, distros, etc, meet all goals, new features, focus on users and their needs, which has always been my focus, that's number one, with me as user number one, and a handful of other people around the world who off and on become priority users in terms of the data they supply me via the debuggers is key to developing features that support that type of hardware.

So it's clear when you talk about debuggers, you're not talking about actual user friendly tools that make getting user data trivially easy,  it's something totally unrelated, which may account for why I basically have no idea what you are talking about? Sometimes it's easy to get blinded by expectations, which then blind you to things that do not match your expectations, thus allowing for the classic missing the forest for the trees?

However, with all this said, I write code that pleases me, I'd be an idiot to waste 10 years of my life writing code that I disliked and found a pain in the butt to work with, that type of code I also do, and that's called commercial code, and you have to pay me a LOT to write it, I mean, a lot. 

But in no sense does this matter, since I do this as a hobby which lets me continuously peer into the disorganized chaos that is human attempts at organization on a slipshod haphazd basis, inxi basically is a living reminder to me of the reality that humans will always fall into chaos if allowed to do so, and inxi will always be this delicate attempt to in near real time, straddle that gap between chaotic and disorganized data as inputs, where basically nothing can be assumed or trusted, even though code has to assume and trust at various levels, and I think that's why you may be slightly confused by what you are seeing in inxi, the problem set inxi deals with is fluid and dynamic and chaotic, and the code has to be agile enough to handle that. I fully agree that in all likelihood most coders will not want to enter into this world, nor should they, you literally could not pay me to do it, it's only because I find the problems and solutions interesting in a certain perverse way that I do it at all. Nobody who does not share this view of the core problems and their solutions that inxi tries to deal with should  ever waste their time on inxi, it will never be that static organized blob of rule following code, it will always be guided by users and their needs, with some users having more priority than others, and the flux of incoherent data that demonstrates the total inability of human practice to self organize in spheres where there are often financial benefits to not doing that in the first place, or where people are too lazy or too incompetent to do it. 

This is the terrain inxi has always followed, I if pressed would view it as a surfer trying to ride a big wave of chaos, but you seem to be looking for a safe pleasure craft where things are controlled and secure. Obviously these two world views can never be reconciled, both have their applications, but the attempt to confuse one for the other would lead to horrible results. I think that may be the mistake you are making, believing there is a one size fits all solution to coding challenges when that is simply not the case. Maybe it's too many years of experience on my part? I have to assume that somehow I did actually manage to learn something over the past decades, and I would further assume, though I've never thought of it in this way before, that inxi is actually the distillation of what I've learned about how to deal with the unstructured chaos that is the reality of global computing data in the real world?  

I've frequently reflected that this chaos is an incredibly accurate display of some key buddhist tenets about how humans confuse manifolds of illusion and delusion for reality... but that's for another world and another life.... but I would not be at all offended if someone were to note that inxi is an attempt at a zen type balancing of chaos and illusion on a strictly technical level. Such attempts are sadly always doomed to failure, because humans have an infinite ability to generate delusion for themselves, but it's still fun trying, lol.

It appears maybe possible that you have not yet mastered the concept of the perfect being the enemy of the in this case very good, that actually takes years to get a hold of, I suspect many coders never learn it, and waste untold years in the process.

I like stepping back and reflecting now and then on some of the meta-principles going on in inxi, it may be helpful to grasp the core principles of inxi, which nothing else out there shares:

* inxi must work on 10 to 12 year old systems, with perl 5.008 or greater installed. Failure to do so is a showstopper bug. I screw this up now and then. While I'm too lazy to do this, a highly valuable test suite is to install all the perl versions and run inxi on each one from 5.008 to current, using the -v8 command to trigger output, possibly with -y1 to make it easier to debug. That's a REAL valuable test for inxi, but I don't feel like doing it because my time is not infinite and usually a user running old stuff will report an oversight in the code to me, which will always be fixed instantly. But that test setup is super easy to do.

* inxi must have extremely flexible output options, and be able to put out output that is user friendly, readable. I use over a decade of usability and web work to determine what I think this should look like, and that was a core reason I did the perl rewrite, I was not able to adequately control output, and could never have done things like output to json or xml in the bash/gawk variant, so that was literally the second feature developed in perl inxi. The first was the user debugger, which is why I find it almost absurd that your belief about what a debugger should be appears to blind you to the absurd levels of debuggers that inxi runs via command line. I will again suggest to you that your belief about what you expect to see is literally blinding you to seeing what inxi actually has in this area.

* inxi must be fast, as fast as possible. This means performance ALWAYS overrides some notion of how to write perl code that has never even been performance tested, or using super slow modules to do something that one single line of code can do 100x faster. The core bottlenecks in inxi are subshell calls, those are EXPENSIVE, nothing else comes remotely close. After my initial optimizations, which I do a bit further now and then when I'm testing a different method (I found a few recently where one way was literally 700x faster than another way). When I find anything that is technically faster, using loop tests up to iterations that wil expose the stuff, and, ideally, running a version of the test on very slow old hardware, where it matters A LOT, because these things can change by entire seconds the time to output. However, one recent optimization did was more psychological than actual total execution time based, I simply moved certain actions to after the first line printed, which sped up the visible appearance of the first line, at which point slight slowdowns became far less visible. These types of optimization tests are something not commonly done because they are slow to do, and often result in mere 100s of thousands of seconds execution differences, but inxi does a LOT of actions so the more of these small optimizations I can achieve (and that's by the way, another area for someone wanting to dip into inxi and challenge their own assumptions about how to use perl can actually do good).

But with this said, the code is now so optimized internally that subshells I believe take at least 90% of the execution time, and I can in most cases do nothing about that. Increasingly I can sometimes replace those subshells with /sys data calls, but I usually have to leave the older methods in place for systems that do not support those /sys data types. But a /sys call is far far faster than a subshell, so you'll notice it's always preferred over anything else.

* inxi must do what its primary users need it to do. Primary users in this case are of course me, first, then distros and their package managers, since they by definition represent their userbase, then serious sys engineers who run serious hardware. Then there's a big dropoff, and regular users appear, first there are regular users who do what I ask and give me the data I need, and who I know from years of dealing with them, and thus trust. Then there's just plain smart people who roughly understand what is going on and how to ask for a feature and give me the info needed to achieve it. Last is regular users who may or may not have valid points, or who ask for things I think are good or bad ideas. The labels show you roughly where an issue request can go.

* Note the position of these last items, it's not a coincidence, the priorities of inxi are roughly in this order. inxi must be easy for me to work on, because the old inxi was hard to work on, and drove me insane. so the way inxi perl is written means its easy for a good coder who can follow logic and doesn't fall into the trap of needing to see things be a certain way, and who can work with large files, and doesn't use bad code editors that fail to make working on large files like this easy or trivial. inxi blew away my expectations in this regard, sure, I add more command line triggered debuggers for certain things I find need or benefit from them, but I have to always look them up, and it's rarely more difficult to just uncomment one for that specific test or issue, then comment it back up than it is to make a dedicated printer for that feature, plus those dedicated lines all have to be parsed by the perl interpreter every time you run inxi, unlike comments, which it knows to ignore.

* now we come to the very last priority of inxi, which actually I spent a lot of time on, inxi is heavily organized internally, it's designed for rapid code folded scrolling [[again, if you find that difficult, get better tools], it's organized in terms of features, internal modules, printers, everything is in sections, and some of those are even documented. And keep in mind, I DO THIS ALL MYSELF!!!!!! nobody else does it, I think you are creating a fantasy world where you actually are confusing large projects that are run in a totally different way with this project, which is run by me, I do the docs, I do the man, I do all the stuff, even though I don't like documentation and rarely keep it up to date. They are notes usually, that's all. But in some cases, they do generally tell you what to use, and do document things like optimizations, values to use in debugging, internal tools.... 

you may simply be from a different planet than I am, I'm not sure, I can't assume, but I don't see many projects who do this much, and certainly not projects run by one person. And when I do, I like them, i3 for example is a shining example of that level of documentation. 

so...... I think maybe you're not the right person here, you seem to want something that inxi will never give you, and that generally is a sign it's the wrong thing for you to really do anything with because you don't seem to actually get the core points of inxi. I ignore most projects out there for exactly this reason, and in fact, it's super, super rare, for me to actually check a project out, because most of them strike me as alien to me and my needs, except as an end user of them, sometimes. But I can say, when I was doing desktop detection development, as I was going through all the window managers, it struck that the only window manager, and I mean the only one, that had fantastic docs, fantastic man, fantastic coding culture, was i3, and that's actually why I now use it on one system, it was literally because I found them to be so superior to anything else I'd looked at while research window managers. Whether or not it uses a specific type of internal debugger that works or doesn't work in a specific way would be last on my list of reasons to contribute or not contribute.

smxi commented:
If it was too long, didn't read, the short version is this: I don't want 'help' that makes my life harder or worse, that's not help, that an obstacle, and i'm only interested in contributors who make my life better, like the guy who was a semi pro editor who went through my entire man page and did a ton of fixes, then the same on the help option menu, and then who dug into the code and did some optimizations. He wasn't a programmer, but he was smart, so that's kind of a come to jesus moment, how can a non programmer who is smart learn without my telling him much how to detect and create a perl optimization who doesn't know perl? it's not a hard thing to figure out, though sometimes someone might not like the obvious answer they get.
